### DISCUSSION TITLE
AuthenticateAsync -> Something went wrong: No token found in payload!

### DISCUSSION THREAD

**Discussion by Robertsmania** [Bukk94/KickLib#2] [Q&A] [answered]
The authentication flow now appears to require a OTP regardless of whether the account has 2FA enabled or not.

Following the suggestion on the discord:
https://discord.com/channels/1047704755511775262/1051232395317235803/1154940917518708766

I hacked in the second token provider request, a **really inconvenient** prompt to enter the one time password they email after the 1st login attempt, and then a second login attempt:

```
    public async Task AuthenticateAsync(string username, string password)
    {
...
        // Call kick-token-provider to get data required for login process
        var tokenProviderResponse = await page.EvaluateFunctionAsync<string>(@"
            async () => {
                const response = await fetch('https://kick.com/kick-token-provider', {
                    method: 'GET',
                    headers: {
                        'Accept': 'application/json',
                        'Content-Type': 'application/json'
                    }
                });

                return response.text();
            }
        ");

        Console.WriteLine("1st Token Provider: " + tokenProviderResponse.ToString());
        
        var tokenProvider = JToken.Parse(tokenProviderResponse);

        // Construct login payload
        var payloadPrep = new ExpandoObject() as IDictionary<string, object>;
        payloadPrep.Add("isMobileRequest", true);
        payloadPrep.Add("email", username);
        payloadPrep.Add("password", password);
        payloadPrep.Add(tokenProvider["nameFieldName"]!.ToString(), "");
        payloadPrep.Add(tokenProvider["validFromFieldName"]!.ToString(), tokenProvider["encryptedValidFrom"]);
        
        var loginPayload = Newtonsoft.Json.JsonConvert.SerializeObject(payloadPrep);
        
        // Using mobile login is easier  
        var loginResponse = await page.EvaluateFunctionAsync<string>($@"
            async () => {{
                const response = await fetch('https://kick.com/mobile/login', {{
                    method: 'POST',
                    headers: {{
                        'Accept': 'application/json',
                        'Content-Type': 'application/json',
                        'X-Xsrf-Token': '{xsrfToken}'
                    }},
                    body: JSON.stringify({loginPayload})
                }});
                return response.text();
            }}
        ");

        Console.WriteLine("1st Login: " + loginResponse.ToString());

        // Call kick-token-provider to get data required for login process
        tokenProviderResponse = await page.EvaluateFunctionAsync<string>(@"
            async () => {
                const response = await fetch('https://kick.com/kick-token-provider', {
                    method: 'GET',
                    headers: {
                        'Accept': 'application/json',
                        'Content-Type': 'application/json'
                    }
                });

                return response.text();
            }
        ");

        Console.WriteLine("2nd Token Provider: " + tokenProviderResponse.ToString());

        Console.WriteLine("Enter OTP:");
        var OTP = Console.ReadLine();

        payloadPrep.Add("one_time_password", OTP);
        loginPayload = Newtonsoft.Json.JsonConvert.SerializeObject(payloadPrep);
        
        // Using mobile login is easier  
        loginResponse = await page.EvaluateFunctionAsync<string>($@"
            async () => {{
                const response = await fetch('https://kick.com/mobile/login', {{
                    method: 'POST',
                    headers: {{
                        'Accept': 'application/json',
                        'Content-Type': 'application/json',
                        'X-Xsrf-Token': '{xsrfToken}'
                    }},
                    body: JSON.stringify({loginPayload})
                }});
                return response.text();
            }}
        ");

        Console.WriteLine("2nd Login: " + loginResponse.ToString());
```

That works, and in initial testing the authentication lasts for the duration of the sessions I've tried.

Obviously having manual intervention to check email and enter a code is undesirable but at least this lets me keep working on stuff.

**Comment by Robertsmania** [comment]
I've gotten around this by changing the account to actually use 2FA, generating my own TOTP and using that.

```
            var secretKey = Base32Encoding.ToBytes(superSekrit);
            var totp = new Totp(secretKey);
            var totpCode = totp.ComputeTotp();
            Console.WriteLine(totpCode);
            
            await api.AuthenticateAsync(sekritEmail, sekritPW, totpCode);

            await client.ListenToChatRoomAsync(2277517);
            await client.ConnectAsync();

            await api.Messages.SendMessageAsync(channelID, "Greetings");
```

I modified AuthenticateAsync to take a third string parameter, the totpCode.

```
public async Task AuthenticateAsync(string username, string password, string totp)
{
...
    // Construct login payload
    var payloadPrep = new ExpandoObject() as IDictionary<string, object>;
    payloadPrep.Add("isMobileRequest", true);
    payloadPrep.Add("email", username);
    payloadPrep.Add("password", password);
    payloadPrep.Add("one_time_password", totp);
    payloadPrep.Add(tokenProvider["nameFieldName"]!.ToString(), "");
    payloadPrep.Add(tokenProvider["validFromFieldName"]!.ToString(), tokenProvider["encryptedValidFrom"]);
    
    var loginPayload = Newtonsoft.Json.JsonConvert.SerializeObject(payloadPrep);
```

It does require that you use the authentication key they provide when you enable 2FA.  

I picked Otp.NET somewhat arbitrarily but it seems like it has many downloads on GitHub and is available both in the full source as well as NuGet.  Your mileage may vary and I have no reason to suggest that library over any others.  Stepping through the code path it uses to generate the key I saw nothing to be concerned about. 
https://github.com/kspearrin/Otp.NET

Here is a video of it in action:
https://www.youtube.com/watch?v=INi12LA3Cyo

**Reply by Bukk94** [reply]
If I get this right, you will enable 2FA on your account but generate own TOTP?
It kinda sucks that it forces you to enable 2FA in anyway. Kick magic 💯 Auth is getting more and more tricky, I hope they will soon release official auth flow.

**Reply by Robertsmania** [reply]
Yes, the way it works for me with no intervention is as follows:

1. Login manually to Kick with the account you want to use, go to Settings / Security.
2. Even if it says 2FA is not enabled, the login process will send you an email for verification, so it really is forced either way and this setting just determines the mechanism.  Click "Enable 2FA" here in the settings to switch from email to secret key / authenticator verification.
3. Next they show you a QR code and a secret key code - COPY AND RECORD THE SECRET KEY.  You should now set it up with an authenticator app like Google Authenticator, Authy, BitWarden, etc.  so you have a mechanism for generating your TOTP and can login manually.  Be sure to record the initial secret key itself.
4. Once you have recoded the secret key and have an authenticator setup, click continue and they require you to provide a 6 digit TOTP.  Enter what you have on your app and it should verify.
5. Now your account is setup to use authenticator TOTP codes rather than email.
6. You might logout and login manually just to verify that it works as expected.

To use this authentication method with KickLib, you will need a procedural way to generate the TOTP codes using the secret key.  All you need to generate a TOTP is the secret key (the login/password is not required).  There are many libraries and systems for computing valid TOTP codes, and for my initial experiments I picked Otp.net:
https://github.com/kspearrin/Otp.NET

```
//Get a byte array for the secret key you got from Kick when enabling 2FA            
var secretKey = Base32Encoding.ToBytes(superSekrit);

//Pass that into the constructor for the Otp.NET Totp class            
var totp = new Totp(secretKey);

//Use the totp object to compute a TOPT code            
var totpCode = totp.ComputeTotp();

//Just to check, I write it to the console so I can see if it matches what I have on the authenticator app. This is not required.            
Console.WriteLine(totpCode);
            
//Pass the login, password and TOPT code to KickLib AuthenticateAsync            
await api.AuthenticateAsync(sekritEmail, sekritPW, totpCode);

//It should succeed, and you can then proceed with authenticated operations
await api.Messages.SendMessageAsync(channelID, "Greetings");
```

There is a small chance it can fail based on time.  Individual TOTP codes are valid in a 30 second window, and the time on various systems is not universally accurate.  In practice, there is almost certainly a time period when the TOPT code you have will be valid based on the time from your system, but invalid on the system at Kick.  So there is a chance you get a code right as it is considered invalid by Kick, try to use it and fail.  So you should consider a retry mechanism to get a new TOTP code and try again in your top level code.

**Reply by Robertsmania** [reply]
With the changes in 5d9285c95a2604f12714de1f32d125e097358016 you can now also provide an empty totp parameter to AuthenticateAsync().

If you do that and provide a valid login/password for an account that does not have secret key/authenticator 2FA enabled, the Kick system will respond to the attempt by sending an email.  You can then make a subsequent login request using the code in the email (those codes are valid for 10 minutes).

The top level code I've been using to test looks like this:

```
            var totpCode = "";
            if (string.IsNullOrEmpty(superSekrit))
            {
                Console.WriteLine("Enter TOTP Code (or return to receive an email):");
                totpCode = Console.ReadLine();
            }
            else
            {
                var secretKey = Base32Encoding.ToBytes(superSekrit);
                var totp = new Totp(secretKey);
                totpCode = totp.ComputeTotp();
                Console.WriteLine(totpCode);
            }

            await api.AuthenticateAsync(sekritEmail, sekritPW, totpCode);

            await api.Messages.SendMessageAsync(channelID, "Greetings");
```

**Reply by Bukk94** [reply]
I'll try to include all of this in Authentication login. I'll test that myself, but based on your very detailed post I'm confident that it will work.

**Reply by Bukk94** [reply, marked as answer]
I implemented TOTP generation inside the `AuthenticationService` so you don't have to bother with OTPs yourself. New auth flow is available in version 1.4.0.
`AuthenticateAsync` method now requires `AuthenticationSettings`, containing Username, Password, and 2FA Authorization code (basically same input as you have here in the example). 
I also updated README with step-by-step guide on how to obtain the code and how to use it.

Really appreciate your feedback and help!
